### PR TITLE
Slack Scheduling

### PR DESCRIPTION
**Pull request by leogrim**
@ryanpbrewster

### CODE REVIEW

File: kifi-backend/modules/shoebox/app/com/keepit/slack/LibraryToSlackChannelPusher.scala [hunk 1]
```diff
@@ -22,13 +22,25 @@ import com.keepit.common.strings._
 import scala.concurrent.{ ExecutionContext, Future }
 import scala.util.{ Failure, Try, Success }
 
+object LibraryToSlackChannelPusher {
+  val maxAcceptableProcessingDuration = Period.minutes(10) // we will wait this long for a process to complete before we assume it is incompetent
+  val defaultDelayBetweenPushes = Period.minutes(30)
+  val minimumDelayFromKeptAt = Period.seconds(20)
```
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/slack/LibraryToSlackChannelPusher.scala]: `FromAddedAt`

File: kifi-backend/modules/shoebox/app/com/keepit/slack/LibraryToSlackChannelPusher.scala [hunk 2]
```diff
@@ -75,9 +83,15 @@ class LibraryToSlackChannelPusherImpl @Inject() (
     }
   }
 
-  def scheduleLibraryToBePushed(libId: Id[Library], when: DateTime)(implicit session: RWSession): Unit = {
+  def pushLibraryPlease(libId: Id[Library]): Unit = db.readWrite { implicit session =>
+    val nextPushAt = clock.now plus minimumDelayFromKeptAt
+    pushLibraryAtLatest(libId, nextPushAt)
+  }
+
+  def pushLibraryAtLatest(libId: Id[Library], when: DateTime)(implicit session: RWSession): Unit = {
```
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/slack/LibraryToSlackChannelPusher.scala:91]: `private def`

File: kifi-backend/modules/shoebox/app/com/keepit/slack/LibraryToSlackChannelPusher.scala [hunk 3]
```diff
@@ -75,9 +83,15 @@ class LibraryToSlackChannelPusherImpl @Inject() (
...
     libToChannelRepo.getActiveByLibrary(libId).foreach { lts =>
-      libToChannelRepo.save(lts.withNextPushAtLatest(when))
+      val updatedLts = lts.withNextPushAtLatest(when)
+      if (updatedLts != lts) libToChannelRepo.save(updatedLts)
```
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/slack/LibraryToSlackChannelPusher.scala:94]: Is this optimization worth it?

File: kifi-backend/modules/shoebox/app/com/keepit/slack/LibraryToSlackChannelPusher.scala [hunk 4]
```diff
@@ -170,25 +184,32 @@ class LibraryToSlackChannelPusherImpl @Inject() (
...
-        val (webhookOpt, ktls, keepsToPush) = db.readOnlyReplica { implicit s =>
+        val (webhookOpt, keepsToPush, lastKtlIdOpt) = db.readOnlyReplica { implicit s =>
           val webhook = slackIncomingWebhookInfoRepo.getForIntegration(lts)
-          val ktls = ktlRepo.getByLibraryFromTimeAndId(libId, lts.lastProcessedAt, lts.lastProcessedKeep)
-          val attributionByKeepId = keepSourceAttributionRepo.getByKeepIds(ktls.map(_.keepId).toSet)
+          val (keeps, lastKtlIdOpt) = {
+            val recentKtls = ktlRepo.getByLibraryAddedAfter(libId, lts.lastProcessedKeep)
+            val keepsByIds = keepRepo.getByIds(recentKtls.map(_.keepId).toSet)
+            val now = clock.now()
+            val stableRecentKeepsAndKtls = recentKtls.flatMap(ktl => keepsByIds.get(ktl.keepId).map(keep => (keep, ktl))).takeWhile {
```
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/slack/LibraryToSlackChannelPusher.scala]: `.map(_ -> ktl)`

File: kifi-backend/modules/shoebox/app/com/keepit/slack/LibraryToSlackChannelPusher.scala [hunk 5]
```diff
@@ -170,25 +184,32 @@ class LibraryToSlackChannelPusherImpl @Inject() (
...
+              case (keep, _) =>
+                val isTitleGoodEnough = keep.title.isDefined || (keep.keptAt isBefore (now minus maxTitleDelayFromKept))
+                val isKeepStable = keep.updatedAt isBefore (now minus delayFromUpdatedAt)
+                (isTitleGoodEnough && isKeepStable) || (keep.keptAt isBefore (now minus maxDelayFromKeptAt))
+            }
+            (stableRecentKeepsAndKtls.map(_._1), stableRecentKeepsAndKtls.lastOption.map(_._2.id.get))
```
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/slack/LibraryToSlackChannelPusher.scala]: `stableRecentKeepsAndKtls.unzip match { case (keeps, ktls) => (keeps, ktls.lastOption.map(_.id.get) }`

File: kifi-backend/modules/shoebox/app/com/keepit/slack/LibraryToSlackChannelPusher.scala [hunk 6]
```diff
@@ -219,7 +240,7 @@ class LibraryToSlackChannelPusherImpl @Inject() (
             libToChannelRepo.save(
               lts
                 .finishedProcessing
-                .withLastProcessedKeep(ktls.lastOption.map(_.id.get))
+                .withLastProcessedKeep(lastKtlIdOpt orElse lts.lastProcessedKeep)
```
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/slack/LibraryToSlackChannelPusher.scala:242]: Ah, clever.

File: kifi-backend/modules/shoebox/app/com/keepit/slack/LibraryToSlackChannelPusher.scala [hunk 7]
```diff
@@ -75,9 +83,15 @@ class LibraryToSlackChannelPusherImpl @Inject() (
     }
   }
 
-  def scheduleLibraryToBePushed(libId: Id[Library], when: DateTime)(implicit session: RWSession): Unit = {
+  def pushLibraryPlease(libId: Id[Library]): Unit = db.readWrite { implicit session =>
+    val nextPushAt = clock.now plus minimumDelayFromKeptAt
+    pushLibraryAtLatest(libId, nextPushAt)
+  }
```
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/slack/LibraryToSlackChannelPusher.scala:89]: This does not actually push to the library. Why change the name from `scheduleToBePushed`?

File: kifi-backend/modules/shoebox/test/com/keepit/slack/LibraryToSlackChannelPusherTest.scala [hunk 8]
```diff
@@ -33,6 +35,11 @@ class LibraryToSlackChannelPusherTest extends TestKitSupport with SpecificationL
     FakeClockModule()
   )
 
+  def pushUpdatesToSlackSurely(libraryId: Id[Library])(implicit injector: Injector): Map[Id[LibraryToSlackChannel], Boolean] = {
+    fakeClock += LibraryToSlackChannelPusher.maxDelayFromKeptAt
+    Await.result(inject[LibraryToSlackChannelPusher].pushUpdatesToSlack(libraryId), Duration.Inf)
+  }
```
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/test/com/keepit/slack/LibraryToSlackChannelPusherTest.scala:41]: :smirk_cat: sure

File: kifi-backend/modules/shoebox/test/com/keepit/slack/LibraryToSlackChannelPusherTest.scala [hunk 9]
```diff
@@ -125,17 +131,16 @@ class LibraryToSlackChannelPusherTest extends TestKitSupport with SpecificationL
...
-          Await.result(inject[LibraryToSlackChannelPusher].pushUpdatesToSlack(lib.id.get), Duration.Inf)
+          pushUpdatesToSlackSurely(lib.id.get)
           slackClient.pushedMessagesByWebhook(webhook.url) must beEmpty
 
           // 2 keeps => 1 msg, 2 lines
           fakeClock += Period.days(1)
           db.readWrite { implicit s =>
             KeepFactory.keeps(2).map(_.withUser(user).withLibrary(lib).withKeptAt(fakeClock.now).withTitle(titles.next())).saved
-            libToSlackPusher.scheduleLibraryToBePushed(lib.id.get, fakeClock.now)
+            libToSlackPusher.pushLibraryAtLatest(lib.id.get, fakeClock.now)
```
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/test/com/keepit/slack/LibraryToSlackChannelPusherTest.scala:141]: Don't think this line is needed anymore.

File: kifi-backend/modules/shoebox/test/com/keepit/slack/LibraryToSlackChannelPusherTest.scala [hunk 10]
```diff
@@ -144,9 +149,9 @@ class LibraryToSlackChannelPusherTest extends TestKitSupport with SpecificationL
           fakeClock += Period.days(1)
           db.readWrite { implicit s =>
             KeepFactory.keeps(20).map(_.withUser(user).withLibrary(lib).withKeptAt(fakeClock.now).withTitle(titles.next())).saved
-            libToSlackPusher.scheduleLibraryToBePushed(lib.id.get, fakeClock.now)
+            libToSlackPusher.pushLibraryAtLatest(lib.id.get, fakeClock.now)
```
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/test/com/keepit/slack/LibraryToSlackChannelPusherTest.scala:152]: same here

File: kifi-backend/modules/shoebox/test/com/keepit/slack/LibraryToSlackChannelPusherTest.scala [hunk 11]
```diff
@@ -197,10 +201,10 @@ class LibraryToSlackChannelPusherTest extends TestKitSupport with SpecificationL
             val lts = LibraryToSlackChannelFactory.lts().withMembership(stm).withLibrary(lib).withChannel("#eng").saved
             KeepFactory.keep().withUser(user).withLibrary(lib).withKeptAt(fakeClock.now).withTitle("Keep Without Note").saved
             KeepFactory.keep().withUser(user).withLibrary(lib).withKeptAt(fakeClock.now).withTitle("Keep With Note").withNote("My [#favorite] keep [#in the world] is this one").saved
-            libToSlackPusher.scheduleLibraryToBePushed(lib.id.get, fakeClock.now)
+            libToSlackPusher.pushLibraryAtLatest(lib.id.get, fakeClock.now)
```
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/test/com/keepit/slack/LibraryToSlackChannelPusherTest.scala:204]: and again

File: kifi-backend/modules/shoebox/app/com/keepit/slack/LibraryToSlackChannelPusher.scala (hunk 1 above)
Comment (leogrim) [inline kifi-backend/modules/shoebox/app/com/keepit/slack/LibraryToSlackChannelPusher.scala]: It's actually from `keptAt`

File: kifi-backend/modules/shoebox/app/com/keepit/slack/LibraryToSlackChannelPusher.scala (hunk 7 above)
Comment (leogrim) [inline kifi-backend/modules/shoebox/app/com/keepit/slack/LibraryToSlackChannelPusher.scala:89]: Hence the `please` but sure I can rename it.

File: kifi-backend/modules/shoebox/app/com/keepit/slack/LibraryToSlackChannelPusher.scala (hunk 2 above)
Comment (leogrim) [inline kifi-backend/modules/shoebox/app/com/keepit/slack/LibraryToSlackChannelPusher.scala:91]: Used in tests

File: kifi-backend/modules/shoebox/app/com/keepit/slack/LibraryToSlackChannelPusher.scala (hunk 3 above)
Comment (leogrim) [inline kifi-backend/modules/shoebox/app/com/keepit/slack/LibraryToSlackChannelPusher.scala:94]: it's cheap

Comment (leogrim) [inline kifi-backend/modules/shoebox/app/com/keepit/slack/LibraryToSlackChannelPusher.scala:94]: and it's kind of hidden in `withNextPushAtLatest` already

File: kifi-backend/modules/shoebox/app/com/keepit/slack/LibraryToSlackChannelPusher.scala (hunk 4 above)
Comment (leogrim) [inline kifi-backend/modules/shoebox/app/com/keepit/slack/LibraryToSlackChannelPusher.scala]: :+1:

File: kifi-backend/modules/shoebox/app/com/keepit/slack/LibraryToSlackChannelPusher.scala (hunk 5 above)
Comment (leogrim) [inline kifi-backend/modules/shoebox/app/com/keepit/slack/LibraryToSlackChannelPusher.scala]: :+1:

File: kifi-backend/modules/shoebox/app/com/keepit/slack/LibraryToSlackChannelPusher.scala (hunk 6 above)
Comment (leogrim) [inline kifi-backend/modules/shoebox/app/com/keepit/slack/LibraryToSlackChannelPusher.scala:242]: Yep that may have caused a bug some day

File: kifi-backend/modules/shoebox/test/com/keepit/slack/LibraryToSlackChannelPusherTest.scala (hunk 9 above)
Comment (leogrim) [inline kifi-backend/modules/shoebox/test/com/keepit/slack/LibraryToSlackChannelPusherTest.scala:141]: We still need to schedule a push don't we?

**Pull request merged**